ibis/backends/tests/test_export.py

```diff
@@ -456,3 +446,130 @@ def test_empty_memtable(backend, con):
     table = ibis.memtable(expected)
     result = con.execute(table)
     backend.assert_frame_equal(result, expected)
+
+
+@pytest.mark.notimpl(["dask", "druid"])
+@pytest.mark.notimpl(
+    ["impala"], raises=AttributeError, reason="missing `fetchmany` on the cursor"
+)
+@pytest.mark.skipif(pa.__version__ < "12.", reason="Pyarrow >= 12 required")
+class TestDataframeInterchange:
+    """Test the __dataframe__ interchange protocol"""
+
+    def test_dataframe_protocol_no_execute(self, con, alltypes, monkeypatch):
+        t = alltypes.select("int_col", "double_col", "string_col")
+        pa_df = t.to_pyarrow().__dataframe__()
+
+        # Patch `con.to_pyarrow` to allow tracking execution
+        to_pyarrow = unittest.mock.Mock(wraps=con.to_pyarrow)
+        monkeypatch.setattr(con, "to_pyarrow", to_pyarrow)
```

Comment:
I don't love patching and mocks, but for the case of counting method calls (to ensure no execution/at most one execution) I think they're a fine solution. I couldn't find another method within ibis that would suffice, but would be happy to switch if there is one.

Reply:
You should be able to use `mocker.spy` here, which avoids having to mess around with mocking directly. Here's an example where we currently use it to test the call count when registering memtables using the snowflake backend:

```
def test_repeated_memtable_registration(simple_con, mocker):
    data = {"key": list("abc"), "value": [[1], [2], [3]]}
    expected = pd.DataFrame(data)
    t = ibis.memtable(data)

    spy = mocker.spy(simple_con, "_register_in_memory_table")

    n = 2

    for _ in range(n):
        tm.assert_frame_equal(simple_con.execute(t), expected)

    # assert that we called _register_in_memory_table exactly n times
    assert spy.call_count == n
```